Allow CPU translation to be specified using command line options.

MaheshRavishankar opened the pull request:
One advantage of using dynamic pass pipelines is that the translation
of the executable to scalar/native code can be done using a pass
pipeline that itself might be specified as a command line option. The
LowerExecutableTranslation now has an options `use-lowering-pipeline`
that can take a list of passes and then lower them to code that can
then be translated to LLVM using the standard LLVM lowering passes.
The lowering pipeline specified as command line only makes sense if
there is just one executable. If there are multiple executables, then
the same pipeline will be used for all of them.

Also clean up some of the default configuration. This is done by
adding a struct `IREE::HAL::TranslateExecutableInfo` which is right
now a C++ struct. This is mostly staging, since it could potentially
be rolled into `hal.executable.target` operation itself.

asaadaldien reviewed (approved):
LGTM! I can't catch any, but can we annotate this PR to run benchmarks to makesure configurations are getting the right values.

MaheshRavishankar commented:
> LGTM! I can't catch any, but can we annotate this PR to run benchmarks to makesure configurations are getting the right values.

Ok, added it. But I dont know what the baseline numbers are. I checked locally and didnt see any regressions.

benvanik commented on iree/compiler/Conversion/CodegenUtils/FunctionUtils.cpp:
```diff
@@ -16,6 +16,17 @@ namespace iree_compiler {
 
 bool isEntryPoint(FuncOp func) { return func.isPublic(); }
 
+FailureOr<FuncOp> getSingleEntryPointFunction(ModuleOp module) {
```
Is this a hard requirement for some reason? I'd really like to not add back the same restrictions we had on the old code path - being able to have multiple functions - especially on the CPU - is really important. It's an artifact of where we do the final-phase linking that there's only single functions at this point in the lowering, but that's not a requirement - someone could come in with a flow.executable with multiple functions. Looking below it seems like you could do the materialization in a loop over all the exported functions?

MaheshRavishankar replied:
No its not a hard requirement. We could handle multiple entry points, etc. I wrote the code first without any such requirements, but then felt like I was just conjecturing about things, without a way to verify my assumptions are correct. So bit of over-indexing on what is "known works path" vs "should work without this as well". If you feel strongly about it, I can go and revisit this and drop this.

asaadaldien commented:
> > LGTM! I can't catch any, but can we annotate this PR to run benchmarks to makesure configurations are getting the right values.
> 
> Ok, added it. But I dont know what the baseline numbers are. I checked locally and didnt see any regressions.

Thanks! I think @antiagainst recently added a way to run baseline vs PR ?

benvanik commented on iree/compiler/Dialect/HAL/IR/LoweringConfig.h:
```diff
@@ -27,6 +27,30 @@
+namespace IREE {
+namespace HAL {
+/// Struct that for a given hal.target.executable defines how it is translated.
+// TODO(ravishankarm): This could also be converted to an attribute on the
```
yeah we should chat about converting this to an attribute - I'll be (hopefully, after 10 other things) getting to the point where we can plumb the attribute through hopefully in a few weeks - if it has to start as just a loose DictionaryAttr or ArrayAttr that's mapped to this that's probably fine but we'd want to get some basic structure on the top level and avoid the duplication into a c++ class asap - otherwise we end up in the classic plumbing case of adding something new requires changing all of this and all backends using it - and versioning gets tricky if you have saved artifacts with a config and then the c++ type changes/etc.

benvanik replied:
(to clarify - fine for now, and I'm happy to help do the work to convert to an attribute when I get to it - mostly just wanted to stick a placeholder here for us to chat about in the future :)

MaheshRavishankar replied:
Yes, I was hoping to chat about it too. Let me know when you want to. (I am mostly open rest of the day). I can make this just a `DictionaryAttr`. As I was looking at this, it seemed like a way we could basically "encode" how to lower a dispatch region. Either say "use this known pipeline" or a string that says "use this sequence of passes". I think that might open up some other "easy hackability using IREE" avenues. I am happy to change this to `DictionaryAttr`.

MaheshRavishankar commented:
cc @ThomasRaoux

MaheshRavishankar commented:
> > > LGTM! I can't catch any, but can we annotate this PR to run benchmarks to makesure configurations are getting the right values.
> > 
> > 
> > Ok, added it. But I dont know what the baseline numbers are. I checked locally and didnt see any regressions.
> 
> Thanks! I think @antiagainst recently added a way to run baseline vs PR ?

I added the `buildkite:benchmark` label (nothing seems to happening right now, but I think I might need to wait here).

MaheshRavishankar commented:
Had a conversation with Ben offline. Some of his comments point to making the `IREE::HAL::TranslateExecutableInfo` an attribute on the `hal.executable.entry_point` (Which was the goal to begin with). Cleaner to have that done as a follow up to this PR. Landing this one.
